pkg/ioutils/concat.go

```diff
@@ -0,0 +1,123 @@
+package ioutils
+
+import (
+	"io"
+	"errors"
+)
+
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+func max(x, y int) int {
+	if x < y {
+		return y
+	}
+	return x
+}
+
+func seekerSize(s io.Seeker) (int64, error) {
+	cur, err := s.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return 0, err
+	}
+
+	size, err := s.Seek(0, io.SeekEnd)
+	if err != nil {
+		return 0, err
+	}
+
+	if _, err := s.Seek(cur, io.SeekStart); err != nil {
+		return 0, err
+	}
+
+	return size, nil
+}
+
+type concatReadSeekCloser struct {
+	a ReadSeekCloser
+	aSize int64
+	b ReadSeekCloser
+	bSize int64
+	off int64
+}
+
+func (self *concatReadSeekCloser) Read(p []byte) (n int, err error) {
+	// if the read starts within a
+	if self.off < self.aSize {
+		if _, err := self.a.Seek(self.off, io.SeekStart); err != nil {
+			return 0, err
+		}
+
+		nA, err := io.ReadFull(self.a, p[:min(len(p), int(self.aSize - self.off))])
```

Comment:
It should be `self.aSize - self.off`. An easy way of confirming that is considering the extreme case where `self.off` is 0. Then it becomes `p[:self.aSize]` which is the full A size